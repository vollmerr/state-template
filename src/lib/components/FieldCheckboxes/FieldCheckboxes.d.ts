import * as React from 'react';

import * as types from '../../utils/types';

export interface FieldCheckboxesProps extends HTMLInputElement {
  /** Use style variant */
  variant?: ''
    | 'primary';

  /** Options to select from */
  options: Array<types.Option>;

  /** aria-label for checkbox group */
  label: String;

  /** Display inline */
  inline: T.bool,
}

declare class FieldCheckboxes extends React.Component<FieldCheckboxesProps, {}> {}

export default FieldCheckboxes;
