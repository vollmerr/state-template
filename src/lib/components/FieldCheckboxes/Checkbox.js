import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

import * as propUtils from '../../utils/propTypes';

// handle changing array value in redux form
export const handleChange = ({ value, values, onChange }) => (event) => {
  const newValues = [...values];
  // first element is blank, remove it
  if (newValues[0] === '') {
    newValues.shift();
  }
  // is checked, add it to array
  if (event.target.checked) {
    newValues.push(value);
  } else {
    // not checked, remove from array
    newValues.splice(newValues.indexOf(value), 1);
  }
  // update in redux form
  return onChange(newValues);
};

// individual checkbox, applies state-template styling
const Checkbox = (props) => {
  const {
    name,
    value,
    option,
    inline,
    variant,
    disabled,
    onChange,
    className,
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  } = props;

  const values = Array.isArray(value) ? value : [value];

  const id = `${name}-${option.value}`;
  const inputProps = {
    id,
    name,
    disabled,
    type: 'checkbox',
    checked: values.includes(option.value),
    onChange: handleChange({ value: option.value, values, onChange }),
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  };

  let cn = classNames([
    'check',
    className,
  ]);

  if (variant) {
    return (
      <label data-test={'field__checkbox'} className={cn} htmlFor={id}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-md btn-block color-white--checked bg-${variant}--checked rounded-0`}>
          {option.label}
        </span>
      </label>
    );
  }

  cn = classNames([
    'p-l-md m-l-0 m-r-md',
    className,
    inline ? 'form-check-inline' : 'form-check',
  ]);

  return (
    <label data-test={'field__checkbox'} className={cn} htmlFor={id}>
      <input className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-checkbox'} aria-hidden>
        <i className={'ca-gov-icon-check-mark'} />
      </div>
      {option.label}
    </label>
  );
};

Checkbox.propTypes = {
  /** Name of field */
  name: T.string.isRequired,

  /** Value of checkbox */
  value: T.oneOfType([
    T.string,
    T.array,
  ]).isRequired,

  /** Option to select */
  option: propUtils.option.isRequired,

  /** Display inline */
  inline: T.bool,

  /** Use style variant */
  variant: T.oneOf([
    '',
    'primary',
  ]),

  /** Disable the input */
  disabled: T.bool,

  /** Label to render for this checkbox */
  label: T.string,

  /** Called when checkbox changes */
  onChange: T.func.isRequired,

  /** Class name to attach to top level label */
  className: T.string,

  /** Accessible indicator for errors existing */
  'aria-invalid': T.string,

  /** Accessible indicator of related information */
  'aria-describedby': T.string,
};

Checkbox.defaultProps = {
  inline: false,
  variant: '',
  disabled: false,
  label: null,
  className: null,
  'aria-invalid': 'false',
  'aria-describedby': null,
};

export default Checkbox;
