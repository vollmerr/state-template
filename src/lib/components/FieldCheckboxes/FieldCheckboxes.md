```jsx
const initialValues = {
  checkbox: ['value2'],
  requiredCheckbox: ['value2'],
  disabledCheckbox: ['value2'],
};

<ExampleForm form={'checkbox'} initialValues={initialValues}>
  <FieldCheckboxes 
    name={'checkbox'}
    label={'Checkbox'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldCheckboxes 
    required
    name={'requiredCheckbox'}
    label={'Required Checkbox'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
  />
  <FieldCheckboxes 
    disabled
    name={'disabledCheckbox'}
    label={'Disabled Checkbox'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
    required
  />
</ExampleForm>
```

### inline

```jsx
const initialValues = {
  inlineCheckbox: 'value2',
};

<ExampleForm form={'inlineCheckbox'} initialValues={initialValues}>
  <FieldCheckboxes 
    inline
    name={'inlineCheckbox'}
    label={'Inline Checkbox'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
      { value: 'value3', label: 'label 3' },
    ]}
  />
</ExampleForm>
```

### primary variant

```jsx
const initialValues = {
  primaryCheckbox: ['value2'],
  requiredPrimaryCheckbox: ['value2'],
  disabledPrimaryCheckbox: ['value2'],
};

<ExampleForm form={'checkboxPrimary'} initialValues={initialValues}>
  <FieldCheckboxes 
    variant={'primary'}
    name={'primaryCheckbox'}
    label={'Primary Checkbox'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldCheckboxes 
    required
    variant={'primary'}
    name={'requiredPrimaryCheckbox'}
    label={'Required Primary Checkbox'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
  />
  <FieldCheckboxes 
    disabled 
    variant={'primary'}
    name={'disabledPrimaryCheckbox'}
    label={'Disabled Primary Checkbox'}
    options={[
      { value: 'value1', label: 'label 1' },
      { value: 'value2', label: 'label 2' },
    ]}
  />
</ExampleForm>
```
