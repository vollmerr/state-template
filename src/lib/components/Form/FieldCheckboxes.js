import React from 'react';
import T from 'prop-types';

import { isEmptyCheck } from '../../utils/validate';
import FieldWrapper from './FieldWrapper';
import withField from './withField';

// individual checkbox, applies state-template styling
const Checkbox = ({
  input, value, label, variant, disabled, // eslint-disable-line react/prop-types
}) => {
  const checked = input.value.toString().indexOf(value) !== -1;
  const onChange = (event) => {
    const newValue = [...input.value];
    if (newValue[0] === '') {
      newValue.shift();
    }
    if (event.target.checked) {
      newValue.push(value);
    } else {
      newValue.splice(newValue.indexOf(value), 1);
    }
    return input.onChange(newValue);
  };
  const inputProps = {
    disabled,
    checked,
    onChange,
    type: 'checkbox',
    name: input.name,
  };

  if (variant) {
    return (
      <label className={'check'}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-md btn-block color-white--checked bg-${variant}--checked rounded-0`}>{label}</span>
      </label>
    );
  }

  return (
    <label className={'form-check-inline p-l-md m-l-0 m-r-md'}>
      <input className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-checkbox'}>
        <i className={'ca-gov-icon-check-mark'} />
      </div>
      {label}
    </label>
  );
};

// group of checkboxes with optional help text and label
// displays error if validation fails
export const FieldCheckboxes = (props) => {
  const {
    input, options, variant, disabled,
  } = props;

  const disabledClass = disabled ? 'disabled' : '';

  return (
    <FieldWrapper {...props}>
      <div className={disabledClass}>
        {
          options.filter(x => !x.hidden).map(option => (
            <Checkbox
              input={input}
              key={option.value}
              variant={variant}
              disabled={disabled}
              {...option}
            />
          ))
        }
      </div>
    </FieldWrapper>
  );
};

FieldCheckboxes.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'primary',
  ]),
  /** Input from redux-form's Field, attaches name, value, etc */
  input: T.object.isRequired,
  /** Options to select from */
  options: T.arrayOf(
    T.shape({
      label: T.string.isRequired,
      value: T.string.isRequired,
    }),
  ).isRequired,
  /** Disable the input */
  disabled: T.bool,
};

FieldCheckboxes.defaultProps = {
  variant: '',
  disabled: false,
};

const withReduxField = withField(isEmptyCheck);
export default withReduxField(FieldCheckboxes);
