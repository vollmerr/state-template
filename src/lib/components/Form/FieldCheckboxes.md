```jsx
<ExampleForm form={'checkbox'}>
  <FieldCheckboxes 
    name={'checkbox'}
    label={'Checkbox'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldCheckboxes 
    required
    name={'requiredCheckbox'}
    label={'Required Checkbox'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
  <FieldCheckboxes 
    disabled
    name={'disabledCheckbox'}
    label={'Disabled Checkbox'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
</ExampleForm>
```

### primary style

```jsx
<ExampleForm form={'checkboxPrimary'}>
  <FieldCheckboxes 
    primary
    name={'primaryCheckbox'}
    label={'Primary Checkbox'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldCheckboxes 
    primary
    required
    name={'requiredPrimaryCheckbox'}
    label={'Required Primary Checkbox'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
  <FieldCheckboxes 
    primary
    disabled
    name={'disabledPrimaryCheckbox'}
    label={'Disabled Primary Checkbox'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
</ExampleForm>
```
