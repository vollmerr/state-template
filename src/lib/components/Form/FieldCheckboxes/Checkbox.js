import React from 'react';
import T from 'prop-types';

// handle changing array value in redux form
export const handleChange = ({ value, values, onChange }) => (event) => {
  const newValues = [...values];
  // first element is blank, remove it
  if (newValues[0] === '') {
    newValues.shift();
  }
  // is checked, add it to array
  if (event.target.checked) {
    newValues.push(value);
  } else {
    // not checked, remove from array
    newValues.splice(newValues.indexOf(value), 1);
  }
  // update in redux form
  return onChange(newValues);
};

// individual checkbox, applies state-template styling
const Checkbox = ({
  input, value, label, variant, disabled,
}) => {
  const values = Array.isArray(input.value) ? input.value : [input.value];
  const checked = values.includes(value);
  const onChange = handleChange({ value, values, onChange: input.onChange });

  const inputProps = {
    disabled,
    checked,
    onChange,
    type: 'checkbox',
    name: input.name,
  };

  if (variant) {
    return (
      <label data-test={'field-checkbox'} className={'check'}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-md btn-block color-white--checked bg-${variant}--checked rounded-0`}>{label}</span>
      </label>
    );
  }

  return (
    <label data-test={'field-checkbox'} className={'form-check-inline p-l-md m-l-0 m-r-md'}>
      <input className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-checkbox'}>
        <i className={'ca-gov-icon-check-mark'} />
      </div>
      {label}
    </label>
  );
};

Checkbox.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'primary',
  ]),
  /** Input from redux-form's Field, attaches name, value, etc  */
  input: T.object.isRequired,
  /** Value of checkbox */
  value: T.string.isRequired,
  /** Label to display for this checkbox */
  label: T.string.isRequired,
  /** Disable the input */
  disabled: T.bool,
};

Checkbox.defaultProps = {
  variant: '',
  disabled: false,
};

export default Checkbox;
