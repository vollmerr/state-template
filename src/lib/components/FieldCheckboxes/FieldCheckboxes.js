import React from 'react';
import T from 'prop-types';

import { isEmptyCheck } from '../../utils/validate';
import * as propUtils from '../../utils/propTypes';
import { withField } from '../Field';

import Checkbox from './Checkbox';

// group of checkboxes
export const FieldCheckboxes = (props) => {
  const {
    options,
    label,
    ...rest
  } = props;

  return (
    <div
      data-test={'field__checkboxes'}
      className={'field__checkboxes'}
      role={'group'}
      aria-label={label}
    >
      {
        options.filter(x => !x.hidden).map(option => (
          <Checkbox
            key={option.value}
            option={option}
            {...rest}
          />
        ))
      }
    </div>
  );
};

FieldCheckboxes.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'primary',
  ]),

  /** Options to select from */
  options: T.arrayOf(propUtils.option).isRequired,

  /** aria-label for checkbox group */
  label: T.string.isRequired,
};

FieldCheckboxes.defaultProps = {
  variant: '',
};

export default withField(isEmptyCheck)(FieldCheckboxes);
