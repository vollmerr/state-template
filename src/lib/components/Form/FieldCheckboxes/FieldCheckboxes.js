import React from 'react';
import T from 'prop-types';

import { isEmptyCheck } from '../../../utils/validate';

import FieldWrapper from '../FieldWrapper';
import withField from '../withField';
import Checkbox from './Checkbox';

// group of checkboxes with optional help text and label
// displays error if validation fails
export const FieldCheckboxes = (props) => {
  const {
    input, options, variant, disabled, ...rest
  } = props;

  return (
    <FieldWrapper data-test={'field-checkboxes'} disabled={disabled} {...rest}>
      <div className={'field-check'}>
        {
          options.filter(x => !x.hidden).map(option => (
            <Checkbox
              input={input}
              key={option.value}
              variant={variant}
              disabled={disabled}
              {...option}
            />
          ))
        }
      </div>
    </FieldWrapper>
  );
};

FieldCheckboxes.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'primary',
  ]),
  /** Input from redux-form's Field, attaches name, value, etc */
  input: T.object.isRequired,
  /** Options to select from */
  options: T.arrayOf(
    T.shape({
      label: T.string.isRequired,
      value: T.string.isRequired,
    }),
  ).isRequired,
  /** Disable the input */
  disabled: T.bool,
};

FieldCheckboxes.defaultProps = {
  variant: '',
  disabled: false,
};

const withReduxField = withField(isEmptyCheck);
export default withReduxField(FieldCheckboxes);
