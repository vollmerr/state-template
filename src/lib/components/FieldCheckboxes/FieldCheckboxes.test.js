import React from 'react';
import { shallow } from 'enzyme';

import Checkbox from './Checkbox';
import { FieldCheckboxes } from './FieldCheckboxes';

const props = {
  'aria-describedby': 'aria-desc-id',
  'aria-invalid': 'false',
  className: 'test-classname',
  disabled: false,
  id: 'test-id',
  inline: false,
  label: 'test-label',
  name: 'test-name',
  onBlur: jest.fn(),
  onChange: jest.fn(),
  onFocus: jest.fn(),
  options: [
    { value: 'value-1', label: 'label-1' },
    { value: 'value-2', label: 'label-2' },
  ],
  value: 'value-1',
  variant: null,
};

let wrapper;
describe('FieldCheckboxes', () => {
  beforeEach(() => {
    wrapper = shallow(<FieldCheckboxes {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all the options as `Checkbox`s', () => {
    expect(wrapper.find(Checkbox).length).toBe(props.options.length);
  });

  it('should hide `hidden` options', () => {
    const options = [
      ...props.options,
      { value: 'value-3', label: 'label-3', hidden: true },
    ];
    wrapper.setProps({ options });
    expect(wrapper.find(Checkbox).length).toBe(props.options.length);
  });
});
