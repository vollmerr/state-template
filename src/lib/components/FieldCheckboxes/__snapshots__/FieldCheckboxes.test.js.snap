// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldCheckboxes should render correctly 1`] = `
<div
  aria-label="test-label"
  className="field__checkboxes"
  data-test="field__checkboxes"
  role="group"
>
  <Checkbox
    aria-describedby={null}
    aria-invalid="false"
    className={null}
    disabled={false}
    inline={false}
    key="value-1"
    label={null}
    name="test-name"
    onChange={[MockFunction]}
    option={
      Object {
        "label": "label-1",
        "value": "value-1",
      }
    }
    value="value-2"
    variant={null}
  />
  <Checkbox
    aria-describedby={null}
    aria-invalid="false"
    className={null}
    disabled={false}
    inline={false}
    key="value-2"
    label={null}
    name="test-name"
    onChange={[MockFunction]}
    option={
      Object {
        "label": "label-2",
        "value": "value-2",
      }
    }
    value="value-2"
    variant={null}
  />
</div>
`;
